src/btree_utils.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017-2018 Parity Technologies
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::mem::size_of;
+
+// Constants from rust's source:
+// https://doc.rust-lang.org/src/alloc/collections/btree/node.rs.html#43-45
+const B: usize = 6;
+const CAPACITY: usize = 2 * B - 1;
+const MIN_LEN_AFTER_SPLIT: usize = B - 1;
+
+/// Estimate the mem size of a btree.
+pub fn mem_size_of_btree<T>(len: u32) -> usize {
+	if len == 0 {
+		return 0;
+	}
+
+	// We try to estimate the size of the `InternalNode` struct from:
+	// https://doc.rust-lang.org/src/alloc/collections/btree/node.rs.html#97
+	// A btree `LeafNode` has 2*B - 1 (K,V) pairs and (usize, u16, u16) overhead.
+	let leaf_node_size = size_of::<(usize, u16, u16, [T; CAPACITY])>();
+	// An `InternalNode` additionally has 2*B `usize` overhead.
+	let internal_node_size = leaf_node_size + size_of::<[usize; 2 * B]>();
+	// A node can contain between B - 1 and 2*B - 1 elements. We assume 2/3 occupancy.
+	let num_nodes = (len as usize).saturating_div((CAPACITY + MIN_LEN_AFTER_SPLIT) * 2 / 3);
+
+	// If the tree has only one node, it's a leaf node.
+	if num_nodes == 0 {
+		return leaf_node_size;
+	}
+	num_nodes.saturating_mul(internal_node_size)
+}
+
+#[cfg(all(test, feature = "unstable-tests"))]
```

Comment:
```suggestion
#[cfg(test)]
#[rustversion::nightly]
```

https://crates.io/crates/rustversion

This way we don't need to expose an feature for this.

Reply:
Nice, but also had to add a build script to make it work, as stated here: https://github.com/dtolnay/rustversion/issues/8

PTAL !